File: 0948-sort-an-array/0948-sort-an-array.cpp
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        long long l = 0, h = nums.size() - 1;
        merge_sort(nums, l, h);
        return nums;
    }   

private:
    void merge_sort(vector<int>& nums, long long l, long long h) {
        if (l >= h) return;
        
        long long m = l + (h - l) / 2;
        merge_sort(nums, l, m);
        merge_sort(nums, m + 1, h);
        combine(nums, l, m, h);
    }

    void combine(vector<int>& nums, long long l, long long m, long long h) {
        long long lx = m - l + 1, ly = h - m;
        vector<int> x(lx), y(ly);
        
        for (long long i = 0; i < lx; i++) x[i] = nums[l + i];
        for (long long i = 0; i < ly; i++) y[i] = nums[m + 1 + i];

        long long px = 0, py = 0, pa = l;
        
        while (px < lx && py < ly) {
            if (x[px] < y[py]) nums[pa++] = x[px++];
            else nums[pa++] = y[py++];
        }

        while (px < lx) nums[pa++] = x[px++];
        while (py < ly) nums[pa++] = y[py++];
    } 
};

File: 0948-sort-an-array/0948-sort-an-array.py
class Solution(object):
    def sortArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        self.merge_sort(nums, 0, len(nums)-1)
        return nums

    def merge_sort(self, nums, l, h):
        if l >= h:
            return
        m = (l + h) // 2
        self.merge_sort(nums, l, m)
        self.merge_sort(nums, m+1, h)
        self.combine(nums, l, m, h)

    def combine(self, nums, l, m, h):
        lx = m - l + 1
        ly = h - m

        left = [0] * lx
        right = [0] * ly

        for i in range(lx):
            left[i] = nums[l + i]
        for i in range(ly):
            right[i] = nums[m + 1 + i]

        i = j = 0
        k = l

        while i < lx and j < ly:
            if left[i] <= right[j]:
                nums[k] = left[i]
                i += 1
            else:
                nums[k] = right[j]
                j += 1
            k += 1

        while i < lx:
            nums[k] = left[i]
            i += 1
            k += 1

        while j < ly:
            nums[k] = right[j]
            j += 1
            k += 1
